File: S3DE/Engine/Graphics/TextureUnits.cs
using S3DE.Engine.Data;
using S3DE.Engine.Graphics.Textures;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace S3DE.Engine.Graphics
{
    public enum TextureUnit
    {
        Null = -1,
        _0,
        _1,
        _2,
        _3,
        _4,
        _5,
        _6,
        _7,
        _8,
        _9,
        _10,
        _11,
        _12,
        _13,
        _14,
        _15,
        _16,
        _17,
        _18,
        _19,
        _20,
        _21,
        _22,
        _23,
        _24,
        _25,
        _26,
        _27,
        _28,
        _29,
        _30,
        _31,
    }

    public static class TextureUnits
    {
        static ITexture[] Textures;
        static int _MaxSupportedTextureUnits;
        static LinkedQueueList<TextureUnit> UnboundTextureUnits, BoundTextureUnits;

        internal static void Initialize()
        {
            UnboundTextureUnits = new LinkedQueueList<TextureUnit>();
            BoundTextureUnits = new LinkedQueueList<TextureUnit>();
            _MaxSupportedTextureUnits = Renderer.TextureUnitCount;
            Textures = new ITexture[_MaxSupportedTextureUnits];

            for (int i = 0; i < _MaxSupportedTextureUnits; i++)
                UnboundTextureUnits.Enqueue((TextureUnit)i);

            Console.WriteLine($"GPU supports {_MaxSupportedTextureUnits} textureunits");
        }

        public static ITexture GetTextureUnit(TextureUnit texUnit)
        {
            if ((int)texUnit + 1 <= _MaxSupportedTextureUnits)
                return Textures[(int)texUnit];
            else
            {
                throw new ArgumentOutOfRangeException($"The GPU only supports {_MaxSupportedTextureUnits} texture units!");
            }
        }

        public static void BindTextureUnit(ITexture tex, TextureUnit texUnit)
        {
            if ((int)texUnit + 1 <= _MaxSupportedTextureUnits)
            {
                ITexture t = Textures[(int)texUnit];
                if (t == null || !t.Compare(tex))
                {
                    if (t == null)
                        UnboundTextureUnits.Remove(texUnit);
                    else
                    {
                        BoundTextureUnits.Remove(texUnit);
                        ITexture t2 = Textures[(int)texUnit];
                        t2.SetIsBound(false, TextureUnit.Null);
                    }

                    Renderer.BindTextureUnit(tex, texUnit);
                    Textures[(int)texUnit] = tex;
                    tex.SetIsBound(true, texUnit);
                    BoundTextureUnits.Enqueue(texUnit);
                }
            }
        }

        public static TextureUnit BindTexture(ITexture tex)
        {
            TextureUnit t;
            if (!tex.IsBound(out t))
            {
                if (UnboundTextureUnits.Count > 0)
                    t = UnboundTextureUnits.Dequeue();
                else
                    t = BoundTextureUnits.Dequeue();

                tex.Bind(t);
            }
            return t;
        }

        public static void BindTextures(params ITexture[] tex)
        {
            //Find multiple free textureunits.
            //or bind over other textureunits.
            //keep track of which ones we've already bound to so we don't overwrite the textures we are trying to bind.
            //Also skip binding if the texture is already bound.
        }

        public static void UnbindTextureUnit(TextureUnit texUnit)
        {
            if ((int)texUnit + 1 <= _MaxSupportedTextureUnits && Textures[(int)texUnit] != null)
            {
                Renderer.UnbindTextureUnit(texUnit);
                Textures[(int)texUnit].SetIsBound(false, TextureUnit.Null);
                Textures[(int)texUnit] = null;
                BoundTextureUnits.Remove(texUnit);
                UnboundTextureUnits.Enqueue(texUnit);
                Renderer.UnbindTextureUnit(texUnit);
            }
        }
    }
}

File: S3DECore/TextureUnits.cpp
#include "Textures.h"

using namespace S3DECore::Collections;

namespace S3DECore {
	namespace Graphics {
		namespace Textures {

			void TextureUnits::InitTextureUnits() {
				if (!isInitialized) {

					availableTextureUnits = S3DECore::Graphics::Renderer::GetParami(GL_MAX_COMBINED_TEXTURE_IMAGE_UNITS);
					//textureUnits = new vector<Texture^>(availableTextureUnits);
					//textureUnits = gcnew System::Collections::Generic::List<Texture^>(availableTextureUnits);

					textureUnits = gcnew cli::array<Texture^>(availableTextureUnits);
					unboundTexUnits = gcnew LinkedQueueList<int>();

					for (int i = 0; i < availableTextureUnits; i++)
						unboundTexUnits->Enqueue(i);

					boundTexUnits = gcnew LinkedQueueList<int>();
					isInitialized = true;
				}
			}

			Texture^ TextureUnits::GetTexture(uint texUnit) {
				return textureUnits[texUnit];
			}

			uint TextureUnits::BindTexture(Texture^ tex) {
				if (tex->IsBound())
					return tex->GetBoundTexUnit();

				return BindTexture(tex, GetAvailableTextureUnit());
			}

			void TextureUnits::UnbindTexture(Texture^ tex) {
				if (tex->IsBound())
					UnbindTexture(tex->GetBoundTexUnit());
			}

			void TextureUnits::UnbindTexture(uint texUnit) {

				Texture^ bt = GetTexture(texUnit);
				
				if (bt != nullptr) {
					Console::WriteLine(String::Format("Unbinding TextureUnit {0}",texUnit));
					bt->SetIsBound(false);
					boundTexUnits->TryRemove(texUnit);
					unboundTexUnits->Enqueue(texUnit);
					SetActiveTextureUnit(texUnit);
					glBindTexture((int)bt->GetTextureType(), NULL);
				}
			}

			uint TextureUnits::BindTexture(Texture^ tex, uint texUnit) {
				if (tex->IsBound())
					return tex->GetBoundTexUnit();

				Texture^ bt = GetTexture(texUnit);
				
				if (bt != nullptr) {
					bt->SetIsBound(false);
					bt->SetBoundTexUnit(0);
					boundTexUnits->TryRemove(texUnit);
				}
				else {
					unboundTexUnits->TryRemove(texUnit);
				}

				SetActiveTextureUnit(texUnit);
				Console::WriteLine(String::Format("Binding texture to TextureUnit {0}", texUnit));
				glBindTexture((int)tex->GetTextureType(), tex->GetIdentifier());
				textureUnits[texUnit] = tex;
				boundTexUnits->Enqueue(texUnit);
				tex->SetIsBound(true);
				tex->SetBoundTexUnit(texUnit);
				return texUnit;
			}

			uint TextureUnits::GetAvailableTextureUnit() {
				if (unboundTexUnits->Count() > 0)
					return unboundTexUnits->Dequeue();
				else {
					return boundTexUnits->Dequeue();
				}
			}

			uint TextureUnits::GetActiveTextureUnit() {
				return activeTextureUnit;
			}

			void TextureUnits::SetActiveTextureUnit(uint texUnit) {
				if (activeTextureUnit != texUnit) {
					activeTextureUnit = texUnit;
					glActiveTexture(GL_TEXTURE0 + texUnit);
				}
			}
		}
	}
}